table/block_based/block.cc

```diff
@@ -659,17 +648,70 @@ void IndexBlockIter::DecodeCurrentValue(uint32_t shared) {
   }
 }
 
-// Binary search in restart array to find the first restart point that
-// is either the last restart point with a key less than target,
-// which means the key of next restart point is larger than target, or
-// the first restart point with a key = target
+template <class TValue>
+void BlockIter<TValue>::ScanAfterBinarySeek(const Slice& target, uint32_t index,
+                                            bool is_index_key_result,
+                                            const Comparator* comp) {
+  // Linear search (within restart block) for first key >= target
+  if (is_index_key_result) {
+    SeekToRestartPoint(index);
+    Next();
+  } else {
+    SeekToRestartPoint(index);
+    uint32_t max_offset;
+    if (index + 1 < num_restarts_) {
+      max_offset = GetRestartPoint(index + 1);
+    } else {
+      max_offset = restarts_;
+    }
+    bool first = true;
+    while (true) {
+      Next();
+      if (!Valid()) {
+        break;
+      }
+      if (first) {
```

Comment:
I don't like the "bool first" idiom here, as the amount of "duplicated" code without it is minimal. I think it hurts readability, and if the compiler doesn't properly optimize, could hurt efficiency with extra conditional branch.

And some of the code is already essentially already duplicated on either size of is_index_key_result.

Reply:
Done.